### DISCUSSION TITLE
How to define custom element types?

### DISCUSSION THREAD

**Discussion by ehehela** [Ferrite-FEM/Ferrite.jl#1051] [Q&A] [answered]
I want to use a locally refined mesh, which consists of 4-node quadrilateral elements, except the transition layer between small and large elements that are made up of multi-node quadrilateral elements (hanging nodes). 

My question is how to define interpolation and quadrature rules for such elements within the framework of Ferrite. 
I am new to Ferrite and have read part of the dev manual. It seems that I need to define a new reference shape, implement new interpolation and develop a new topology method for the non-conforming mesh.

This seems like a big job. What is the first step to getting started? Are there any dedicated guidance or references available? Any suggestions would be greatly appreciated.

**Comment by termi-official** [comment, marked as answer]
> This seems like a big job.

Yes.

> Are there any dedicated guidance or references available?

Not yet.

> What is the first step to getting started?

If you want the full beta experience I recommend to experiment with https://github.com/Ferrite-FEM/Ferrite.jl/pull/780 . Quadrilaterals are likely safe to use. For 3D we are still tracking down some issues. Also, we have no real debug layer yet, so tracking down issues is still very manual. However, we are almost on the finish line with this PR to support AMR with quad and hex elements.

> It seems that I need to define a new reference shape, implement new interpolation and develop a new topology method for the non-conforming mesh.

Depends on what exactly you want to do. We are also investigating other approaches than p4est to local refinement in parallel, which work directly on basis-function and dof management level.

If you have some simple topology or refinement structure, than you can probably get away with tracking the hanging nodes manually and constructing matrices to project the problem from the non-conforming into a conforming space. https://doi.org/10.1137/18M1193992 explains how to construct the matrices in detail (for the serial case).

**Reply by ehehela** [reply]
Thanks for your answer. However, I have decided to program my own codes because I want to implement XFEM, which may not be easy to achieve in Ferrite. Because I am unsure how to handle the extra DOFs from XFEM in Ferrite. Developing codes based on the versatile and robust Ferrite seems complicated for me as a newcomer.  I hope this is not a bad decision or direction.

**Comment by lijas** [comment]
You don't have to redefine new refshapes or interpolations. You will still use RefCube and Lagrage{2,RefCube,1} as usual.
The difference is that you will have to constrain the hanging nodes in some way. This is typically done with linear constraints, read https://www.dealii.org/current/doxygen/deal.II/step_6.html for a nice explanation. Luckily, we have [linear(affine) constraints in ferrite](https://github.com/Ferrite-FEM/Ferrite.jl/blob/241d181490511e67fcb5d723b67ecb1c22cbe363/src/Dofs/ConstraintHandler.jl#L72).

The tricky part is to find all hanging nodes and create constraints for them. There have been work on this in #780 like Dennis said, but I dont know how easy it is to utilise that.

Another approach would be to have triangle elements in the transition region... in that way you avoid hanging nodes.

**Reply by termi-official** [reply]
> Luckily, we have [linear(affine) constraints in ferrite](https://github.com/Ferrite-FEM/Ferrite.jl/blob/241d181490511e67fcb5d723b67ecb1c22cbe363/src/Dofs/ConstraintHandler.jl#L72). The tricky part is to find all hanging nodes and create constraints for them. 
> 

I just want to add that we do not support nested (or recursive) constraints yet.

**Reply by ehehela** [reply]
Thank you for your reply. However, the hanging nodes are not addressed through constraints in my implementation. Instead, they are however attached to transition elements, which pose special shape functions. The standard 4-node elements are just a particular case when the node count is 4. (See [Ref.](https://doi.org/10.12989/sem.2012.43.3.349)). Therefore, new refshape is needed, I think.

**Comment by KnutAM** [comment]
> Therefore, new refshape is needed, I think.

Yes, that would be needed for the pentagon elements that you are asking about. Adding a new reference shape should be quite easy and also defining the interpolations. I assume that you might need some special treatment for the geometrical interpolation for concave pentagons, but if you can avoid that, I think it is quite straight forward and we can help you to get there. 

> extra DOFs from XFEM in Ferrite.

I'm not super-familiar with XFEM. But I think that should be feasible, although it will be harder than adding pentagon elements. I think you (a) need to define some non-conventional interpolations and (b) pass some extra data regarding the location of the enrichment during cell values `reinit!` 

So if you go your own way, at I'd be happy to hear how it goes! Maybe we can integrate it with Ferrite in the future. And in any case welcome to the julia finite element community!
